drivers/sensor/grow_r502a/grow_r502a.c

```diff
@@ -0,0 +1,713 @@
+/*
+ * Copyright (c) 2021 Linumiz
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT hzgrow_grow_r502a
+
+#include <init.h>
+#include <kernel.h>
+#include <drivers/gpio.h>
+#include <drivers/sensor.h>
+#include <drivers/uart.h>
+
+#include "grow_r502a.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(GROW_R502A, CONFIG_SENSOR_LOG_LEVEL);
+
+static void send_packet(const struct device *dev, uint8_t pkg_type,
+		uint8_t *pkg_data, uint16_t len)
+{
+	uint8_t buffer[64];
+	uint16_t check_sum, pkg_len;
+
+	pkg_len = len + FPS_CHECKSUM_LEN;
+	check_sum = pkg_len + pkg_type;
+
+	buffer[0] = FPS_STARTCODE >> 8;
+	buffer[1] = FPS_STARTCODE & 0xFF;
+	buffer[2] = (uint8_t)(FPS_DEFAULT_ADDRESS >> 24);
+	buffer[3] = (uint8_t)(FPS_DEFAULT_ADDRESS >> 16);
+	buffer[4] = (uint8_t)(FPS_DEFAULT_ADDRESS >> 8);
+	buffer[5] = (uint8_t)(FPS_DEFAULT_ADDRESS & 0xFF);
+	buffer[6] = pkg_type;
+	buffer[7] = pkg_len >> 8;
+	buffer[8] = pkg_len & 0xFF;
+	for (int i = FPS_HEADER_LEN ; i < (len + FPS_HEADER_LEN) ; i++) {
+		buffer[i] = pkg_data[i-FPS_HEADER_LEN];
+		check_sum += buffer[i];
+	}
+	buffer[len + FPS_HEADER_LEN] = check_sum >> 8;
+	buffer[len + FPS_HEADER_LEN + 1] = check_sum & 0xFF;
+
+	for (uint8_t i = 0 ; i < (pkg_len + FPS_HEADER_LEN) ; i++) {
+		uart_poll_out(dev, buffer[i]);
+	}
+
+}
+
+static int receive_data(const struct device *dev, uint8_t *buffer)
+{
+	int i = 0;
+	uint8_t buf_len = 0;
+	uint64_t end = sys_clock_timeout_end_calc(K_MSEC(DEFAULT_TIMEOUT));
+
+	while (1) {
+		uint8_t c;
+
+		int64_t remaining = end - sys_clock_tick_get();
+
+		if (remaining <= 0) {
+			return -EIO;
+		}
+
+		if (uart_poll_in(dev, &c) == 0) {
+			buffer[i++] = c;
+			if (i == FPS_HEADER_LEN) {
+				buf_len = FPS_HEADER_LEN + buffer[FPS_HEADER_LEN - 1];
+			}
+			if (i == buf_len) {
+				break;
+			}
+		}
+	}
+
+	for (int i = 0 ; i < buf_len ; i++) {
+		LOG_DBG("0x%X ", buffer[i]);
+	}
+
+	return 0;
+}
+
+static uint8_t fps_led_control(const struct device *dev, struct led_params *led_control)
+{
+	struct grow_r502a_data *drv_data = dev->data;
+	uint8_t data[5] = {FPS_LED_CONFIG, led_control->color_code, led_control->speed,
+		led_control->color, led_control->cycle};
+
+	send_packet(dev, FPS_COMMAND_PACKET, data, sizeof(data));
+
+	int ret = receive_data(dev, drv_data->buffer);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (drv_data->buffer[6] != FPS_ACK_PACKET) {
+		LOG_ERR("Error receiving ack packet 0x%X", drv_data->buffer[6]);
+		return -EIO;
+	}
+
+	if (drv_data->buffer[9] == FPS_OK) {
```

Comment:
this and others: if certain buffer indexes contain named fields, do something like this:

```c
#define MY_FIELD 9

if (drv_data->buffer[MY_FIELD] == FPS_OK) {
```